scan/lib/scan/runner.rb

```diff
@@ -135,27 +137,118 @@ def retryable_tests(input)
-      if Scan.config[:disable_xcpretty]
-        unless tests_exit_status == 0
-          UI.test_failure!("Test execution failed. Exit status: #{tests_exit_status}")
+    def find_filename(type)
+      index = Scan.config[:output_types].split(',').index(type)
+      return nil if index.nil?
+      return (Scan.config[:output_files] || "").split(',')[index]
+    end
+
+    def output_html?
+      return Scan.config[:output_types].split(',').include?('html')
+    end
+
+    def output_junit?
+      return Scan.config[:output_types].split(',').include?('junit')
+    end
+
+    def output_json_compilation_database?
+      return Scan.config[:output_types].split(',').include?('json-compilation-database')
+    end
+
+    def output_html_filename
+      return find_filename('html')
+    end
+
+    def output_junit_filename
+      return find_filename('junit')
+    end
+
+    def output_json_compilation_database_filename
+      return find_filename('json-compilation-database')
+    end
+
+    def trainer_test_results
+      require "trainer"
+
+      results = {
+        number_of_tests: 0,
+        number_of_failures: 0,
+        number_of_retries: 0
+      }
+
+      result_bundle_path = Scan.cache[:result_bundle_path]
+
+      output_path = Scan.config[:output_directory] || Dir.mktmpdir
+      output_path = File.absolute_path(output_path)
+
+      UI.crash!("A -resultBundlePath is needed to parse the test results. This should not have happened. Please file an issue.") unless result_bundle_path
+
+      params = {
+        path: result_bundle_path,
+        silent: true
+      }
+
+      formatter = Scan.config[:xcodebuild_formatter].chomp
+
+      if output_html? && formatter != 'xcpretty'
+        UI.error("HTML output is only available with `xcodebuild_formatter: 'xcpretty'` right now...")
+      end
+
+      if output_json_compilation_database? && formatter != 'xcpretty'
+        UI.error("JSON Compilation Database output is only available with `xcodebuild_formatter: 'xcpretty'` right now...")
+      end
+
+      if output_junit?
+        if formatter == 'xcpretty'
+          UI.verbose("Generating junit report with xcpretty")
+        else
+          UI.verbose("Generating junit report with trainer")
+          params[:output_filename] = output_junit_filename || "report.junit"
+          params[:output_directory] = output_path
         end
-        return
       end
 
-      result = TestResultParser.new.parse_result(test_results)
-      SlackPoster.new.run(result)
+      resulting_paths = Trainer::TestParser.auto_convert(params)
+      resulting_paths.each do |path, data|
+        results[:number_of_tests] += data[:number_of_tests_excluding_retries]
+        results[:number_of_failures] += data[:number_of_failures_excluding_retries]
+        results[:number_of_retries] += data[:number_of_retries]
+      end
+
+      return results
+    end
+
+    def handle_results(tests_exit_status)
+      results = trainer_test_results
```

Comment:
this isn't reached for me when any tests fail, because `test_app` in this file is raising an exception instead of returning an exit status, and this existing code doesn't handle exception:
```
    def run
      handle_results(test_app)
```
if it did then I do get a fulfilled test results table and report.junit file, for example:
```
    def run
      begin
        results = test_app
      rescue => ex
      ensure
        handle_results(results || -1)
        raise ex unless ex.nil?
```
but I guess we'd want to fix `test_app` instead

Reply:
note though that not raising an exception anymore may be a breaking change for some... I myself wrap `run_tests` with `begin` (so far based on empirical observations) so my `trainer` call runs for failed builds too (well, that's when we most want it, right 😄)

Reply:
@gobetti Ohhhhh, this makes so much sense now!! Can you try setting `fail_build: false` and see if that "fixes" it for you? 😇

I totally see how this existing fail logic doesn't make sense anymore. I will look into! Thanks so much for all the detail 🙌

Reply:
ha! glad you asked, because I considered making that a **d.** case in my huge post - no, adding `fail_build: false` doesn't make a difference. But given I can't find `fail_build` referenced at all in scan/runner.rb I didn't know that was worth consideration

Reply:
I see now where `fail_build` is used in fastlane's code but it's _outside_ to the `run` method I shared above, so it makes sense that test_app still raises and consequently skips handle_results/trainer_test_results

also fyi, not caused by this PR but affects the desired behavior here - although my failure was just tests failing (exit status 65) the exception is coming as a FastlaneBuildFailure in which case `fail_build` is ignored in the current code:
```
        rescue FastlaneCore::Interface::FastlaneBuildFailure => ex
          # Specifically catching FastlaneBuildFailure to prevent build/compile errors from being
          # silenced when :fail_build is set to false
          # :fail_build should only suppress testing failures
          raise ex
```

Reply:
the build failure is defined in scan/error_handler.rb:
```
        when /Testing failed/
          UI.build_failure!("Error building the application. #{details}")
```
if I'm interpreting correctly, it sounds inconsistent to the code block I pasted on my previous message - `:fail_build should only suppress testing failures` seems to suggest a testing failure shouldn't be a FastlaneBuildFailure

Reply:
okkk I think I've narrowed it down 🙌 the logic in `scan/error_handler.rb` was apparently tailored for xcpretty and still works well if I force `xcodebuild_formatter: 'xcpretty'` - however if I use xcbeautify or `xcodebuild_formatter: ''`, then the `output` that is used by `handle_build_error` is something that doesn't work with the logic currently in place

Reply:
https://github.com/fastlane/fastlane/issues/19827